consensus/core/src/round_tracker.rs

```diff
@@ -0,0 +1,519 @@
+                .round_tracker_current_accepted_round_gaps
+                .with_label_values(&[&authority.hostname])
+                .set(last_proposed_round as i64 - *low as i64);
+        }
+
+        // TODO: consider using own quorum round gap to control proposing in addition to
+        // propagation delay. For now they seem to be about the same.
+
+        // It is possible more blocks arrive at a quorum of peers before the get_latest_rounds
+        // requests arrive.
+        // Using the lower bound to increase sensitivity about block propagation issues
+        // that can reduce round rate.
+        // Because of the nature of TCP and block streaming, propagation delay is expected to be
+        // 0 in most cases, even when the actual latency of broadcasting blocks is high.
+        // We will use the min propagation delay from either accepted or received rounds.
+        // As stated above new blocks can arrive after the rounds have been probed, so its
+        // likely accepted rounds from new blocks will provide us with the more accurate
+        // propagation delay which is important because we now calculate the propagation
+        // delay more frequently then before.
+        let propagation_delay = last_proposed_round
+            .saturating_sub(received_quorum_rounds[own_index].0)
+            .min(last_proposed_round.saturating_sub(accepted_quorum_rounds[own_index].0));
+
+        node_metrics
+            .round_tracker_propagation_delays
+            .observe(propagation_delay as f64);
+        node_metrics
+            .round_tracker_last_propagation_delay
+            .set(propagation_delay as i64);
+
+        debug!(
+            "Computed propagation delay of {propagation_delay} based on last proposed \
+                round ({last_proposed_round})."
+        );
+
+        propagation_delay
+    }
+
+    // Returns received quorum rounds & accepted quorum rounds
+    pub(crate) fn calculate_quorum_rounds(&self) -> (Vec<QuorumRound>, Vec<QuorumRound>) {
+        let received_quorum_rounds = self.compute_received_quorum_rounds();
+        let accepted_quorum_rounds = self.compute_accepted_quorum_rounds();
+
+        debug!(
+            "Computed received quorum round per authority: {}",
+            self.context
+                .committee
+                .authorities()
+                .zip(received_quorum_rounds.iter())
+                .map(|((i, _), rounds)| format!("{i}: {rounds:?}"))
+                .join(", ")
+        );
+        debug!(
+            "Computed accepted quorum round per authority: {}",
+            self.context
+                .committee
+                .authorities()
+                .zip(accepted_quorum_rounds.iter())
+                .map(|((i, _), rounds)| format!("{i}: {rounds:?}"))
+                .join(", ")
+        );
+
+        (received_quorum_rounds, accepted_quorum_rounds)
+    }
+
+    fn compute_received_quorum_rounds(&self) -> Vec<QuorumRound> {
```

Comment:
Maybe expose compute_received_quorum_rounds() and compute_accepted_quorum_rounds() separately, and only call the latter in ancestor selection?